### ISSUE TITLE
Refactor and Re-Organize MLNs

### ISSUE THREAD

**Issue by LinguList** [lingpy/lingpy#267]
We should definitely change the MLNs:
1. divide graphic output and algorithmic analyses
2. use a different method for ASR
3. use different handling of ASR (but which?)
4. allow for externally calculated ASR
5. refine the workflow

Maybe there are more things, let's collect them here.

**Comment by Anaphory** [comment]
What inputs does the MLN algorithm take, again?

Did I get this roughly right:
 - a tree
 - a thing mapping feature to a set of values for every node (internals and leaves), which we get from Wordlist + ASR
 - weights for various things not conforming to the tree
 - some scalar to restrict the number of reconstructed synonyms

?

**Comment by LinguList** [comment]
okay, the question is what you WANT. And this is the current problem, as it leaves no choice.

1. assemble data which needs to be in terms of *phyletic patterns*, that is, presence or absence of a given feature, but the feature should ideally be something like a cognate set, as homoplastic features are problematic for the borrowing assumption, as they can always independently arise (like the typology stuff, you know), this is currently automatically created from a word list, I'll say more on it below
2. calculate or provide a reference tree in some way
3. carry out a gain-loss mapping or in simpler terms, an ancestral state reconstruction on the patterns using the reference tree
4. calculate the ancestral states of each internal node in the tree, given the inferred scenarios (the backwards procedure in parsimony)
5. identify the best of different patterns by comparing recent with ancestral vocabulary distributions, using the kruskal-wallis test
6. identify homoplasies by using the best model and display them as weighted network, try to resolve edges of three homoplastic nodes on the tree (three origins) using minimal spanning trees
6. plot the tree with the superimposted edges on the network
7. re-calculate a geographic network by using only extant taxa and transposing internal edges on them using geographic distance measures and convex hulls
8. plot the spatial network using basemap

You are completely right: these steps do not need to rely on each other, but I guess you also understand that people might wan to repeat them in a flush. A solution for MLN in the future might be to keep the current code as is for historical reasons, potentially putting it on some cemetery for old algos which are more or less deprecated. At the same time using the new [newick library](https://github.com/glottobank/python-newick) to handle trees in lingpy (an old idea), and a new parsimony implementation which is already there, but currently only as some "plugin" or "addon" published in two publications along with the code. For historical reasons, re-implementation of some of the even simpler algorithms or baselines for ASR can also be included, so one could think of an ancestral-state-reconstruction module, rather than MLN.

Tree drawing and network plotting should be made an external module, with some specified input format, like the tree, the edges, their weight, probably even not more, probably even in GLM format, as this is the format that is currently also used and quite suitable for networks. 

Calculation of ancestral state distributions could be easily super-imposed by expanding the ASR module to allow for some useful output, and treatment of the nodes in the tree.

Calculation of the MLN is actually a graph-procedure, some kind of a projection or the reduction of nodes and the like, so it could be handled along with the calculation of spatial network creation. 

I think, the core is the ASR module, which currently already uses (almost) the new newick library. Splitting plotting and calculating should also be rather straightforward, as long as the formats for passing data to file and back are clear.

**Comment by Anaphory** [comment]
> I guess you also understand that people might wan to repeat them in a flush.

Sure! As I said somewhere, wrapping function into objects which cache intermediate results and provide defaults is good, in particular when calculations are as expensive as ours often are.

Your plan looks good to me (but I know neither how what things currently happen internally, nor have I properly dug into the articles defining MLN and variants).

**Comment by LinguList** [comment]
well, I think I have some rather immediate motivation of adding the ASR
code in some form, but not before beginning of next year.

Many things are straightforward. A basic question, however, is the
representation of ASR results, which is probably causing most of the
confusion in current MLN, as there is not really straightforward way
known to me to do it. Having some idea for a minimal format, showing the
results of an ASR analysis for a given character (or also many
characters) is what is needed. Currently, the format is working, but
rather ad-hoc, and has no counterparts in any literature.
